fix: handle relative path glob raw import, fix #7307

toshify opened the pull request:
### Description

`import.meta.glob("../index.html", { as: "raw" })` fails to resolve the correct path during a `readFile` operation.

This fix will add a test for the failing use case of glob-importing with both a relative parent folder path and a raw import,
and fix this by using the unmanipulated file path as reference when reading the file. `file` is manipulated to add the `../`-parts from the glob pattern, while `files[i]` is the bare path relative to `base`.

Closes #7307

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [X] Bug fix
- [ ] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [X] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [X] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [X] Ideally, include relevant tests that fail without this PR but pass with it.

toshify committed:
fix: handle relative path glob raw import, fix #7307

bluwy commented on packages/playground/glob-import/index.html:
```diff
@@ -52,3 +53,22 @@
+<script type="module">
+  // issue #7307
+  const relativeRawModules = import.meta.globEager(
+    '../dynamic-import/nested/shar*.js',
```
I think it would be safer if we can glob a file within the `glob-import` playground. Otherwise a change in the `dynamic-import` playground could cause unexpected test failures for future contributors.

toshify replied:
I'm ok with limiting the test to `glob-import`, but it makes the test a tad contrite, since the purpose is to test directory tree traversal above the parent folder which in this case is `glob-import` itself. Navigating back into `glob-import` would be a use-case we wouldn't expect in real life. However, the test is probably robust enough, and at the very least a reminder to evaluate the possibility.

bluwy commented on packages/playground/glob-import/index.html:
```diff
@@ -52,3 +53,22 @@
 </script>
+
+<script type="module">
+  // issue #7307
```
A micro nitpick from me 😬 I don't think we need to put an issue number here (and in other parts of the code in this PR) as this should initially just work, but seems like an edge case we didn't catch. The Git history should have this PR linked too.

toshify replied:
Good feedback, unnecessary comments should be removed.

toshify committed:
fix: handle relative path glob raw import, fix #7307

Contain test within `glob-import`-folder
Remove comments referencing issue number
